enterprise/cmd/llm-proxy/internal/httpapi/upstream.go

```diff
@@ -0,0 +1,128 @@
+package httpapi
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/sourcegraph/log"
+
+	"github.com/sourcegraph/sourcegraph/enterprise/cmd/llm-proxy/internal/actor"
+	"github.com/sourcegraph/sourcegraph/enterprise/cmd/llm-proxy/internal/events"
+	"github.com/sourcegraph/sourcegraph/enterprise/cmd/llm-proxy/internal/response"
+	llmproxy "github.com/sourcegraph/sourcegraph/enterprise/internal/llm-proxy"
+	"github.com/sourcegraph/sourcegraph/internal/httpcli"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+type bodyTransformer[T any] func(*T)
+type requestTransformer func(*http.Request)
+type requestMetadataRetriever[T any] func(T) (promptCharacterCount int, model string, additionalMetadata map[string]any)
+type responseParser[T any] func(T, io.Reader) (completionCharacterCount int)
+
+func makeUpstreamHandler[T any](logger log.Logger, eventLogger events.Logger, upstreamAPIURL string, bodyTrans bodyTransformer[T], rmr requestMetadataRetriever[T], reqTrans requestTransformer, respParser responseParser[T]) http.Handler {
```

Comment:
nit - it kind of looked like there is supposed to be 2 type parameters, one for request and one for response, but we only request a concrete request type in the parameters, it would be clearer if we were more explicit with the type parameters i.e.

```suggestion
type bodyTransformer[RequestT any] func(*RequestT)
type requestTransformer func(*http.Request)
type requestMetadataRetriever[RequestT any] func(RequestT) (promptCharacterCount int, model string, additionalMetadata map[string]any)
type responseParser[RequestT any] func(req RequestT, resp io.Reader) (completionCharacterCount int)

func makeUpstreamHandler[RequestT any](logger log.Logger, eventLogger events.Logger, upstreamAPIURL string, bodyTrans bodyTransformer[RequestT], rmr requestMetadataRetriever[RequestT], reqTrans requestTransformer, respParser responseParser[RequestT]) http.Handler {
```

Reply:
another nit - would `upstreamReverseProxy` be a more appropriate name for this construct?

Reply:
Renamed the type, I think upstreamReverseProxy is a bit loose since it does a bit more than that with all the logging and such, so I went with the generic Handler term. No strong opinion though 😀